evm/src/cpu/control_flow.rs

```diff
@@ -91,13 +96,11 @@ pub fn eval_packed_generic<P: PackedField>(
     yield_constr.constraint_transition(is_last_noncpu_cycle * (nv.is_kernel_mode - P::ONES));
     yield_constr.constraint_transition(is_last_noncpu_cycle * nv.stack_len);
 
-    // The last row must be a CPU cycle row.
-    yield_constr.constraint_last_row(is_cpu_cycle - P::ONES);
-    // Also, the last row's `program_counter` must be inside the `halt` infinite loop. Note that
-    // that loop consists of two instructions, so we must check for `halt` and `halt_inner` labels.
-    let (halt_pc0, halt_pc1) = get_halt_pcs::<P::Scalar>();
-    yield_constr
-        .constraint_last_row((lv.program_counter - halt_pc0) * (lv.program_counter - halt_pc1));
+    // The last row must be a CPU cycle row or a dummy padding row.
+    yield_constr.constraint_last_row((is_cpu_cycle - P::ONES) * (lv.null - P::ONES));
+    // Also, the last row's `program_counter` must be inside the `halt` infinite loop.
+    let halt_pc = get_halt_pc::<P::Scalar>();
+    yield_constr.constraint_last_row(lv.program_counter - halt_pc);
```

Comment:
This doesn't make sense anymore. I would make sure that if lv.null is set, then program counter is at the `halt` instructions.
```suggestion
    // Also, a padding row's `program_counter` must be at the `halt` label.
    // In particular, it ensures that the first padding row may only be added
    // after we jumped to the `halt` function. Subsequent padding rows may set
    // the `program_counter` to arbitrary values (there's no transition
    // constraints) so we can place this requirement on them too.
    let halt_pc = get_halt_pc::<P::Scalar>();
    yield_constr.constraint(lv.null * (lv.program_counter - halt_pc));
```